File: src/subsystems/navigation/aruco_bt/src/aruco_node.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <sensor_msgs/msg/camera_info.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <opencv2/aruco.hpp>
#include <tf2_ros/transform_listener.h>
#include <tf2_ros/buffer.h>
#include <tf2_geometry_msgs/tf2_geometry_msgs.hpp>

class ArUcoNode : public rclcpp::Node
{
public:
    ArUcoNode() : Node("aruco_node")
    {
        this->declare_parameter("marker_size", 0.2);
        
        bool sim = this->get_parameter("use_sim_time").as_bool();
        marker_size_ = this->get_parameter("marker_size").as_double();

        if (sim) {
            marker_size_ = 0.50;
        }
        
        std::string image_topic = "/zed/zed_node/left/image_rect_color";
        std::string camera_info_topic = "/zed/zed_node/left/camera_info";

        RCLCPP_INFO(this->get_logger(), "sim=%s", sim ? "true" : "false");
        RCLCPP_INFO(this->get_logger(), "marker_size=%.3f meters", marker_size_);
        RCLCPP_INFO(this->get_logger(), "Subscribing to image feed: %s", image_topic.c_str());
        RCLCPP_INFO(this->get_logger(), "Subscribing to camera info: %s", camera_info_topic.c_str());

        aruco_dict_ = cv::aruco::getPredefinedDictionary(cv::aruco::DICT_4X4_50);
        aruco_params_ = cv::aruco::DetectorParameters::create();

        tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
        tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);

        image_sub_ = this->create_subscription<sensor_msgs::msg::Image>(
            image_topic, 10,
            std::bind(&ArUcoNode::imageCallback, this, std::placeholders::_1));
        
        camera_info_sub_ = this->create_subscription<sensor_msgs::msg::CameraInfo>(
            camera_info_topic, 10,
            std::bind(&ArUcoNode::cameraInfoCallback, this, std::placeholders::_1));

        image_pub_ = this->create_publisher<sensor_msgs::msg::Image>("/aruco_annotated_img", 10);
        pose_pub_ = this->create_publisher<geometry_msgs::msg::PoseStamped>("/aruco_pose", 10);

        marker_id_ = -1;
        camera_info_received_ = false;
    }

private:
    void cameraInfoCallback(const sensor_msgs::msg::CameraInfo::SharedPtr msg)
    {
        if (!camera_info_received_) {
            // Extract camera matrix (K) and distortion coefficients
            camera_matrix_ = cv::Mat(3, 3, CV_64F);
            for (int i = 0; i < 9; i++) {
                camera_matrix_.at<double>(i / 3, i % 3) = msg->k[i];
            }

            dist_coeffs_ = cv::Mat(msg->d.size(), 1, CV_64F);
            for (size_t i = 0; i < msg->d.size(); i++) {
                dist_coeffs_.at<double>(i) = msg->d[i];
            }

            camera_info_received_ = true;
            RCLCPP_INFO(this->get_logger(), "Camera info received");
            camera_info_sub_.reset();
        }
    }

    void imageCallback(const sensor_msgs::msg::Image::SharedPtr msg)
    {
        // Convert ROS image to OpenCV format
        cv_bridge::CvImagePtr cv_ptr;
        try {
            cv_ptr = cv_bridge::toCvCopy(msg, "mono8");
        } catch (cv_bridge::Exception& e) {
            RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
            return;
        }

        // Convert grayscale to color for annotation
        cv::Mat frame_color;
        cv::cvtColor(cv_ptr->image, frame_color, cv::COLOR_GRAY2BGR);

        // Detect ArUco markers
        detectArucoMarkers(cv_ptr->image);

        // If markers detected, annotate the image and estimate pose
        if (marker_id_ >= 0 && !corners_.empty()) {
            // Draw bounding box around marker
            for (size_t i = 0; i < 4; i++) {
                cv::Point start_point(static_cast<int>(corners_[i].x), 
                                     static_cast<int>(corners_[i].y));
                cv::Point end_point(static_cast<int>(corners_[(i + 1) % 4].x), 
                                   static_cast<int>(corners_[(i + 1) % 4].y));
                cv::line(frame_color, start_point, end_point, cv::Scalar(0, 255, 0), 2);
            }

            // Add text label with marker ID
            cv::Point text_pos(static_cast<int>(corners_[0].x), 
                              static_cast<int>(corners_[0].y) - 20);
            std::string label = "id: " + std::to_string(marker_id_);
            cv::putText(frame_color, label, text_pos, 
                       cv::FONT_HERSHEY_SIMPLEX, 0.7, 
                       cv::Scalar(0, 0, 255), 2, cv::LINE_AA);

            // Estimate pose if camera info is available
            if (camera_info_received_) {
                estimateAndPublishPose(msg);
                
                // Draw axis on the image to visualize pose
                cv::aruco::drawAxis(frame_color, camera_matrix_, dist_coeffs_, 
                                   rvec_, tvec_, marker_size_ * 0.5);
            }
        }

        // Convert back to ROS message and publish
        sensor_msgs::msg::Image::SharedPtr annotated_msg = 
            cv_bridge::CvImage(msg->header, "bgr8", frame_color).toImageMsg();
        image_pub_->publish(*annotated_msg);
    }

    void detectArucoMarkers(const cv::Mat& frame)
    {
        std::vector<std::vector<cv::Point2f>> corners;
        std::vector<int> ids;

        cv::aruco::detectMarkers(frame, aruco_dict_, corners, ids, aruco_params_);

        if (!ids.empty()) {
            // Store all corners for the first detected marker
            marker_id_ = ids[0];
            all_corners_.clear();
            all_corners_.push_back(corners[0]);
            
            corners_.clear();
            for (const auto& pt : corners[0]) {
                corners_.push_back(pt);
            }
        } else {
            marker_id_ = -1;
            corners_.clear();
            all_corners_.clear();
        }
    }

    void estimateAndPublishPose(const sensor_msgs::msg::Image::SharedPtr& msg)
    {
        if (all_corners_.empty()) {
            return;
        }

        if (marker_size_ <= 0) {
            RCLCPP_WARN(this->get_logger(), "Invalid marker_size_: %.3f", marker_size_);
            return;
        }

        std::vector<cv::Vec3d> rvecs, tvecs;
        
        // Estimate pose of the marker
        cv::aruco::estimatePoseSingleMarkers(all_corners_, marker_size_, 
                                            camera_matrix_, dist_coeffs_, 
                                            rvecs, tvecs);

        // Store for visualization
        rvec_ = rvecs[0];
        tvec_ = tvecs[0];

        // Convert rotation vector to rotation matrix, then to quaternion
        cv::Mat rotation_matrix;
        cv::Rodrigues(rvec_, rotation_matrix);
        rotation_matrix.convertTo(rotation_matrix, CV_64F);

        // Convert rotation matrix to quaternion (tf2)
        tf2::Matrix3x3 m(
          rotation_matrix.at<double>(0,0), rotation_matrix.at<double>(0,1), rotation_matrix.at<double>(0,2),
          rotation_matrix.at<double>(1,0), rotation_matrix.at<double>(1,1), rotation_matrix.at<double>(1,2),
          rotation_matrix.at<double>(2,0), rotation_matrix.at<double>(2,1), rotation_matrix.at<double>(2,2)
        );
        tf2::Quaternion q;
        m.getRotation(q);
        q.normalize();

        // Create PoseStamped in camera frame
        geometry_msgs::msg::PoseStamped pose_camera;
        pose_camera.header.stamp = msg->header.stamp;
        pose_camera.header.frame_id = msg->header.frame_id;

        // Position (translation vector from camera to marker)
        pose_camera.pose.position.x = tvec_[0];
        pose_camera.pose.position.y = tvec_[1];
        pose_camera.pose.position.z = tvec_[2];

        // Orientation (quaternion)
        pose_camera.pose.orientation.x = 0.0;
        pose_camera.pose.orientation.y = 0.0;
        pose_camera.pose.orientation.z = 0.0;
        pose_camera.pose.orientation.w = 1.0;


        // Transform to map frame
        try {
            geometry_msgs::msg::PoseStamped pose_map;
            tf_buffer_->transform(pose_camera, pose_map, "map", tf2::durationFromSec(0.1));

            geometry_msgs::msg::TransformStamped tf_map_base =
            tf_buffer_->lookupTransform("map", "base_footprint", msg->header.stamp, tf2::durationFromSec(0.1));
            
            // Overwrite goal orientation with rover's current orientation
            pose_map.pose.orientation.x = tf_map_base.transform.rotation.x;
            pose_map.pose.orientation.y = tf_map_base.transform.rotation.y;
            pose_map.pose.orientation.z = tf_map_base.transform.rotation.z;
            pose_map.pose.orientation.w = tf_map_base.transform.rotation.w;

            pose_pub_->publish(pose_map);
        } catch (tf2::TransformException& ex) {
            RCLCPP_WARN(this->get_logger(), "Could not transform to map frame: %s", ex.what());
        }
    }

    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr image_sub_;
    rclcpp::Subscription<sensor_msgs::msg::CameraInfo>::SharedPtr camera_info_sub_;
    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr image_pub_;
    rclcpp::Publisher<geometry_msgs::msg::PoseStamped>::SharedPtr pose_pub_;

    std::shared_ptr<tf2_ros::Buffer> tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;

    cv::Ptr<cv::aruco::Dictionary> aruco_dict_;
    cv::Ptr<cv::aruco::DetectorParameters> aruco_params_;

    cv::Mat camera_matrix_;
    cv::Mat dist_coeffs_;
    bool camera_info_received_;
    double marker_size_;

    int marker_id_;
    std::vector<cv::Point2f> corners_;
    std::vector<std::vector<cv::Point2f>> all_corners_;
    
    cv::Vec3d rvec_;  // Rotation vector
    cv::Vec3d tvec_;  // Translation vector
};

int main(int argc, char** argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<ArUcoNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: src/subsystems/navigation/aruco_bt/aruco_bt/aruco_node.py
import numpy as np
import cv2
import cv2.aruco

import rclpy
from rclpy.node import Node
from rclpy.duration import Duration
from rclpy.time import Time
from sensor_msgs.msg import Image, CameraInfo
from geometry_msgs.msg import PoseStamped
from vision_msgs.msg import Detection2D, BoundingBox2D, ObjectHypothesisWithPose
from cv_bridge import CvBridge
import tf2_ros
import tf2_geometry_msgs


class ArUcoNode(Node):
    def __init__(self):
        super().__init__('aruco_node')

        self.declare_parameter('marker_size', 0.2)
        self.declare_parameter('image_topic', '/zed/zed_node/left/image_rect_color')
        self.declare_parameter('camera_info_topic', '/zed/zed_node/left/camera_info')

        sim = self.get_parameter('use_sim_time').get_parameter_value().bool_value
        self.marker_size_ = self.get_parameter('marker_size').get_parameter_value().double_value
        image_topic = self.get_parameter('image_topic').get_parameter_value().string_value
        camera_info_topic = self.get_parameter('camera_info_topic').get_parameter_value().string_value

        if sim:
            self.marker_size_ = 0.50

        self.get_logger().info(f'sim={"true" if sim else "false"}')
        self.get_logger().info(f'marker_size={self.marker_size_:.3f} meters')
        self.get_logger().info(f'Subscribing to image feed: {image_topic}')
        self.get_logger().info(f'Subscribing to camera info: {camera_info_topic}')

        aruco_dict = cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_4X4_50)
        aruco_params = cv2.aruco.DetectorParameters()
        self.aruco_detector_ = cv2.aruco.ArucoDetector(aruco_dict, aruco_params)

        self.tf_buffer_ = tf2_ros.Buffer()
        self.tf_listener_ = tf2_ros.TransformListener(self.tf_buffer_, self)

        self.bridge_ = CvBridge()

        self.image_sub_ = self.create_subscription(
            Image, image_topic, self.image_callback, 1)

        self.camera_info_sub_ = self.create_subscription(
            CameraInfo, camera_info_topic, self.camera_info_callback, 10)

        self.detection_pub_ = self.create_publisher(Detection2D, '/aruco_detection', 10)

        self.marker_id_ = -1
        self.corners_ = []
        self.all_corners_ = []
        self.camera_info_received_ = False
        self.camera_matrix_ = None
        self.dist_coeffs_ = None
        self.rvec_ = None
        self.tvec_ = None

    def camera_info_callback(self, msg):
        if not self.camera_info_received_:
            self.camera_matrix_ = np.array(msg.k, dtype=np.float64).reshape(3, 3)
            self.dist_coeffs_ = np.array(msg.d, dtype=np.float64).reshape(-1, 1)
            self.camera_info_received_ = True
            self.get_logger().info('Camera info received')
            self.destroy_subscription(self.camera_info_sub_)
            self.camera_info_sub_ = None

    def image_callback(self, msg):
        try:
            cv_ptr = self.bridge_.imgmsg_to_cv2(msg, 'mono8')
        except Exception as e:
            self.get_logger().error(f'cv_bridge exception: {e}')
            return

        self.detect_aruco_markers(cv_ptr)

        if self.marker_id_ >= 0 and len(self.corners_) > 0 and self.camera_info_received_:
            pose = self.estimate_pose(msg)
            if pose is not None:
                xs = [pt[0] for pt in self.corners_]
                ys = [pt[1] for pt in self.corners_]

                detection = Detection2D()
                detection.header.stamp = self.get_clock().now().to_msg()
                detection.header.frame_id = 'map'
                detection.bbox.center.position.x = float(sum(xs) / 4)
                detection.bbox.center.position.y = float(sum(ys) / 4)
                detection.bbox.size_x = float(max(xs) - min(xs))
                detection.bbox.size_y = float(max(ys) - min(ys))

                hypothesis = ObjectHypothesisWithPose()
                hypothesis.hypothesis.class_id = str(self.marker_id_)
                hypothesis.hypothesis.score = 1.0
                hypothesis.pose.pose = pose
                detection.results.append(hypothesis)

                self.detection_pub_.publish(detection)

    def detect_aruco_markers(self, frame):
        corners, ids, _ = self.aruco_detector_.detectMarkers(frame)

        if ids is not None and len(ids) > 0:
            self.marker_id_ = int(ids[0][0])
            self.all_corners_ = [corners[0]]
            self.corners_ = [pt for pt in corners[0][0]]
        else:
            self.marker_id_ = -1
            self.corners_ = []
            self.all_corners_ = []

    def estimate_pose(self, msg):
        if len(self.all_corners_) == 0:
            return None

        if self.marker_size_ <= 0:
            self.get_logger().warn(f'Invalid marker_size_: {self.marker_size_:.3f}')
            return None

        half = self.marker_size_ / 2.0
        obj_points = np.array([[-half,  half, 0],[ half,  half, 0],[ half, -half, 0],[-half, -half, 0]], dtype=np.float32)

        img_points = self.all_corners_[0].reshape(4, 2)
        _, self.rvec_, self.tvec_ = cv2.solvePnP(
            obj_points, img_points,
            self.camera_matrix_, self.dist_coeffs_,
            flags=cv2.SOLVEPNP_IPPE_SQUARE)

        pose_camera = PoseStamped()
        pose_camera.header.frame_id = msg.header.frame_id

        pose_camera.pose.position.x = float(self.tvec_[0][0])
        pose_camera.pose.position.y = float(self.tvec_[1][0])
        pose_camera.pose.position.z = float(self.tvec_[2][0])

        pose_camera.pose.orientation.x = 0.0
        pose_camera.pose.orientation.y = 0.0
        pose_camera.pose.orientation.z = 0.0
        pose_camera.pose.orientation.w = 1.0

        try:
            pose_camera.header.stamp = Time().to_msg()
            pose_map = self.tf_buffer_.transform(pose_camera, 'map', timeout=Duration(seconds=0.1))

            tf_map_base = self.tf_buffer_.lookup_transform(
                'map', 'base_footprint', Time(),
                timeout=Duration(seconds=0.1))

            pose_map.pose.orientation.x = tf_map_base.transform.rotation.x
            pose_map.pose.orientation.y = tf_map_base.transform.rotation.y
            pose_map.pose.orientation.z = tf_map_base.transform.rotation.z
            pose_map.pose.orientation.w = tf_map_base.transform.rotation.w

            return pose_map.pose
        except Exception as ex:
            self.get_logger().warn(f'Could not transform to map frame: {ex}')
            return None


def main(args=None):
    rclpy.init(args=args)
    node = ArUcoNode()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()